File: Pieces/Bishop.mjs
// import Square from "../Square.mjs";
import Piece from "./Piece.mjs";

export default class Bishop extends Piece {
  constructor(isWhite) {
    super(isWhite);
  }

  pieceSymbol = () => {
    return this.isWhite ? "♝" : "♗";
  };

  move = (board, initialSquare, destinationSquare) => {
    if (destinationSquare.piece?.isWhite === this.isWhite) return false;
    if (!this.isDiagonalMove(initialSquare, destinationSquare)) return false;

    return this.isValidPath(board, initialSquare, destinationSquare);
  };

  isValidPath = (initialSquare, destinationSquare) => {
    let stepX = destinationSquare.x > initialSquare.x ? 1 : -1;
    let stepY = destinationSquare.y > initialSquare.y ? 1 : -1;

    let x = initialSquare.x + stepX;
    let y = initialSquare.y + stepY;

    while (x !== destinationSquare.x && y !== destinationSquare.y) {
      if (board[x][y].piece) {
        return false; // There is a piece blocking the path
      }
      x += stepX;
      y += stepY;
    }

    return true;
  };

  isDiagonalMove = (initialSquare, destinationSquare) => {
    let dx = Math.abs(initialSquare.x - destinationSquare.x);
    let dy = Math.abs(initialSquare.y - destinationSquare.y);

    return dx === dy;
  };
}

File: Pieces/py/Bishop.py
import Piece

class Bishop(Piece):
    def __init__(self, is_white) -> None:
        super().__init__(is_white)

    def move(self, board, initial_square, destination_square):
        if destination_square.piece and destination_square.piece.is_white == self.is_white:
            return False
        
        if not self.is_diagonal_move(board, initial_square, destination_square):
            return False
        
        return self.is_valid_path(board, initial_square, destination_square)
        
    def is_valid_path(self, board, initial_square, destination_square):
        step_x = 1 if destination_square.x > initial_square.x else -1
        step_y = 1 if destination_square.y > initial_square.y else -1

        x = initial_square.x + step_x
        y = initial_square.y + step_y

        while x != destination_square.x and y != destination_square.y:
            if board[x][y].piece:
                return False
            x += step_x
            y += step_y

        return True

    def is_diagonal_move(self, initial_square, destination_square):
        dx = abs(initial_square.x - destination_square.x)
        dy = abs(initial_square.y - destination_square.y)

        return dx == dy